### DISCUSSION TITLE
Introducing: Localize your keywords!

### DISCUSSION THREAD

**Discussion by TiBiBa** [hedyorg/hedy#2080] [Announcements]
As the frequent users might have already notice on alpha we have been working on something big: Localized keywords! It was already possible to create Hedy programs using English keywords and localized keywords (if supported) to make learning programming easier and more fun. You were even able to change languages with the press of one button!

But with our latest update you will be able to set a _preferred keyword language_ in your profile settings. The level texts, starting programs and adventures will automatically be shown with your set keyword language. This makes is even easier to start your Hedy adventure. Currently there is already support for Dutch, English, French, Spanish and Arabic.

Feel free to translate more languages and reach out to us if you need any help! Go check it out!

**Comment by OnnoEbbens** [comment]
Thanks for creating this new feature! For my classes (1st year havo/vwo) I will stick to the English keywords and the Dutch language but it is nice to have the option in Hedy. Also the implementation is really well done so it was not difficult for my students to choose the languages. The only thing I don't like so much is the language switcher (NL/EN) on the top right of the code screen in Hedy (see screenshot below). For me it is just an extra button that I don't need and therefore an unnecessary distraction in front of the code.

![image](https://user-images.githubusercontent.com/10562903/158570291-2701655c-d43b-4fe5-bc39-0ba2279e137b.png)

I don't know if many people are using this button? Would it be possible to hide it as a teacher for my students?

**Reply by TiBiBa** [reply]
Thanks for the feedback! We discussed this as well (@Felienne and I) if and when this switcher should be visible. If I remember correctly we decided to always make it visible when a language is chosen that also supports a keyword language. Which is also the case for dutch. 

One solution might be to add this as a class customizations option for teachers? Similar to the current "mandatory developer's mode" and "hide cheatsheet" options. Let me know! I can make some time to work on this today / tomorrow so we can deploy this quickly!

**Reply by Felienne** [reply]
Yeah I think this is what we meant with #2009 maybe? That the teacher can enable/disable the switcher? 

We could also imagine that kids can disable it themselves in their profile setting. It can be handy to have it in view if you have a kid that uses a language that the teacher does not know, and it can be helpful to gradually learn English commands. But it is true it is "yet another thing" in view.

Maybe one easy option to improve the layout in any case is to move the button to the bottom of Ace so it does not block the code so often?

**Reply by TiBiBa** [reply]
I can easily add a hide function to the class customization. However, students will have to pay attention themselves that they set the current keyword language in their profile. It might be an issue if the teacher is explaining everything with English keywords whereas the student accidentally has it set to dutch (and sees all examples with dutch keywords).

Shortly stated: Hiding the switcher with class customizations is easy to implement. Enforce a profile keyword language through customizations is hard to implement.

**Reply by OnnoEbbens** [reply]
For me hiding the switcher would be enough. So doing it from class customization would be nice. I don't need to enforce a profile keyword language as a teacher.

**Reply by OnnoEbbens** [reply]
In the last days some students noticed something confusing. In case you use Dutch with English commands and you use a Dutch word in your code that is coincidentally the translation of an English command, the word is highlighted in red as if it were a command. This happens more frequently than I initially thought. As an example, see the fortune adventure in level 4 (https://www.hedycode.com/hedy/4/#fortune) with the code `vraag is ask 'Wat wil je weten?'`. Would it be doable to implement an English-command-only option where users cannot switch to Dutch and no highlighting of Dutch translations of English commands?

**Reply by TiBiBa** [reply]
It should be possible to only allow the English commands. However, the currently described problem is actual intended as a feature. If I remember correctly the goal was to always allow students to use English keywords as well as the keywords of their own language. Even within the same program. So in this case the students will always be allowed to use dutch keywords as well. The selected keyword language is only used to parse the keyword placeholders in the examples, level-defaults and adventures with the chosen language but the actual parser used is still on the user language (and not the user keyword language).

I'm not sure what would be the best approach to solve this issue. We could simply approach it like "if the English keyword language is chosen, also use the English syntax highlighting and parser". This should be easy to implement. However, as I recall this current situation was chosen on purpose. What do you think? @Felienne
